shared-bindings/aes/aes.c

```diff
@@ -0,0 +1,355 @@
+      {MP_QSTR_counter, MP_ARG_OBJ},
+      {MP_QSTR_segment_size, MP_ARG_INT},
+  };
+  mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+
+  // Set defaults. These will be overridden in mp_arg_parse_all() if an
+  // argument is provided.
+  args[ARG_mode].u_int = 0;
+  args[ARG_IV].u_obj = NULL;
+  args[ARG_counter].u_int = 0;
+  args[ARG_segment_size].u_int =
+      8; // Only useful in CFB mode, which we don't support
+  mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args),
+                   allowed_args, args);
+
+  aes_obj_t *self = m_new_obj(aes_obj_t);
+  self->base.type = &aes_aes_type;
+
+  mp_buffer_info_t bufinfo;
+
+  const uint8_t *key = NULL;
+  uint32_t key_length = 0;
+  if (mp_get_buffer(args[ARG_key].u_obj, &bufinfo, MP_BUFFER_READ)) {
+    if ((bufinfo.len != 16) && (bufinfo.len != 24) && (bufinfo.len != 32)) {
+      mp_raise_TypeError(translate("Key must be 16, 24, or 32 bytes long"));
+    }
+    key = bufinfo.buf;
+    key_length = bufinfo.len;
+  } else {
+    mp_raise_TypeError(translate("No key was specified"));
+  }
+
+  int mode = args[ARG_mode].u_int;
+  switch (args[ARG_mode].u_int) {
+  case AES_MODE_CBC: /* CBC */
+    break;
+  case AES_MODE_ECB: /* ECB */
+    break;
+  case AES_MODE_CTR: /* CTR */
+    break;
+  case 0:
+    mode = AES_MODE_ECB;
+    break;
+  default:
+    mp_raise_TypeError(translate("Requested AES mode is unsupported"));
+  }
+
+  // IV is required for CBC mode and is ignored for other modes.
+  const uint8_t *iv = NULL;
+  if (args[ARG_IV].u_obj != NULL &&
+      mp_get_buffer(args[ARG_IV].u_obj, &bufinfo, MP_BUFFER_READ)) {
+    if (bufinfo.len != AES_BLOCKLEN) {
+      mp_raise_TypeError_varg(translate("IV must be %d bytes long"),
+                              AES_BLOCKLEN);
+    }
+    iv = bufinfo.buf;
+  }
+
+  common_hal_aes_construct(self, key, key_length, iv, mode,
+                           args[ARG_counter].u_int);
+  return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC mp_obj_t aes_rekey(size_t n_args, const mp_obj_t *pos_args) {
+  aes_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+
+  size_t key_length = 0;
+  const uint8_t *key = (const uint8_t *)mp_obj_str_get_data(pos_args[1], &key_length);
+  if (key == NULL) {
+    mp_raise_ValueError(translate("No key was specified"));
+  }
+  if ((key_length != 16) && (key_length != 24) && (key_length != 32)) {
+    mp_raise_TypeError(translate("Key must be 16, 24, or 32 bytes long"));
+  }
+
+  const uint8_t *iv = NULL;
+  if (n_args > 2) {
+    size_t iv_length = 0;
+    iv = (const uint8_t *)mp_obj_str_get_data(pos_args[2], &iv_length);
+    if (iv_length != AES_BLOCKLEN) {
+      mp_raise_TypeError_varg(translate("IV must be %d bytes long"),
+                              AES_BLOCKLEN);
+    }
+  }
+
+  common_hal_aes_rekey(self, key, key_length, iv);
+  return mp_const_none;
+}
+
+MP_DEFINE_CONST_FUN_OBJ_VAR(aes_rekey_obj, 2, aes_rekey);
+
+
+STATIC mp_obj_t aes_set_mode(mp_obj_t aes_obj, mp_obj_t mode_obj) {
```

Comment:
Please add mode as a property instead of having a `set_mode` function. Here is an example: https://github.com/adafruit/circuitpython/blob/master/shared-bindings/busio/SPI.c#L372-L389